Final steps for 6.2c

flomnes opened the pull request:
- Fix glitches with `OptimEntityContainer`
- One `OptimEntityContainer` for each problem
- Write master.mps and 1-1.mps
- Write structure in `std::cout`

TODO
- [ ] Write structure into structure.txt

flomnes committed:
Update src/solver/optim-model-filler/ComponentFiller.cpp

flomnes committed:
Use std::ranges::any_of

flomnes committed:
Add MasterAndSubPbVariables::write

flomnes committed:
Modeler logic

flomnes committed:
Skip constraints & objectives in master

flomnes committed:
Filter vars in OptimEntityContainer::addFromSystemComponents

flomnes committed:
Write master.mps, 1-1.mps in modeler mode

guilpier-code commented on src/modeler/modeler/Modeler.cpp:
```diff
@@ -144,38 +142,52 @@ void Modeler::run() const
-                                                     data.scenarioGroupRepository,
-                                                     masterAndSubPbvars.get());
+    // Subproblem
+    SystemLinearProblemBuilder subproblem_builder(data.system.get(),
```
I would expect that we first build + dump the **master problem**, and then build + dump+ solve the **sub-problem**.

flomnes replied:
All the same ! I can do the master problem first

flomnes committed:
Fix additional variables

guilpier-code commented on src/solver/optim-model-filler/ComponentFiller.cpp:
```diff
@@ -104,11 +104,22 @@ void MasterAndSubPbVariables::add(const std::vector<std::string>& names, unsigne
+        }
+    }
+}
+
```
Note that line (line # 295): 
```c++
const auto& evaluationContext = optimEntityContainer_.getEvaluationContext(component_);
```
is useless

flomnes replied:
Yes, I removed it on develop and can remove it here as well to avoid conflicts

flomnes committed:
Fix logic

flomnes committed:
Reserve + comments

guilpier-code commented on src/solver/optim-model-filler/ComponentFiller.cpp:
```diff
@@ -334,7 +345,8 @@ void ComponentFiller::addVariables(const LinearProblemApi::FillContext& ctx)
         }
 
-        if (variable.location() == Modeler::Config::Location::MASTER_AND_SUBPROBLEMS
+        if ((variable.location() == Modeler::Config::Location::MASTER_AND_SUBPROBLEMS
```
Why do we also add variables belonging to **MASTER** problem only to **masterAndSubPbvars_** ?
They seem to have nothing to do with content of ile **structure.txt**

flomnes replied:
You're absolutely right, I will make the change

guilpier-code commented on src/modeler-optimisation-container/OptimEntityContainer.cpp:
```diff
@@ -47,13 +48,21 @@ void OptimEntityContainer::addFromSystemComponents(const std::vector<Component>&
+                modelVariableGlobalIndices.push_back(variableGlobalIndex);
+                ++variableGlobalIndex;
+            }
+            else
+            {
+                // We add dummy data to avoid "holes" in modelVariableGlobalIndices
+                // but these are not to be used
+                modelVariableGlobalIndices.push_back(-1);
+            }
```
Why do we need to add `-1`s ? 
Does it even work considering what is done in **ComponentFiller::addVariables(...)** regarding variables counts ?
I mean that, in this (previous) member function : 
```c++
optimEntityContainer_.addStartColumn();
```
is called only for variables that are "compatible" with the problem type.

flomnes replied:
We add dummy indices here to avoid index shifts (filtering)

guilpier-code commented on src/optimisation/linear-problem-api/include/antares/optimisation/linear-problem-api/linearProblem.h:
```diff
@@ -86,7 +86,7 @@ class ILinearProblem
     virtual IMipSolution* solve(bool verboseSolver) = 0;
 
-    virtual void WriteLP(const std::string& filename) const = 0;
+    virtual void Write(const std::string& filename) const = 0;
```
This member function **Write** does not have to be part of **ILinearProblem** interface class : what if we want to write a linear problem in another way ? Would we add a new write member function to the interface ?
We should lighten this interface class by removing this function, and for class  `OrtoolsLinearProblem`, this **Write** function could be a free function : 
```c++
void writeMPS(const OrtoolsLinearProblem& problem, const std::string & filename);
```

flomnes replied:
done

guilpier-code commented on src/optimisation/linear-problem-mpsolver-impl/include/antares/optimisation/linear-problem-mpsolver-impl/linearProblem.h:
```diff
@@ -79,7 +79,7 @@ class OrtoolsLinearProblem: public LinearProblemApi::ILinearProblem
 
     OrtoolsMipSolution* solve(bool verboseSolver) override;
-    void WriteLP(const std::string& filename) const override;
+    void Write(const std::string& filename) const override;
```
see above comment on parent interface class

flomnes replied:
done

guilpier-code commented on src/solver/optim-model-filler/ComponentFiller.cpp:
About filtering **variables** or **objectives**, we should consider using piping with a filter code style.
That's the reason why it was introduced in C++20, as you well know, and we naturally use "filter" when we talk about it.

flomnes replied:
Yes that can be done

guilpier-code replied:
We could even define the new **locationFilter** at construction of **ComponentFiller**. Same for objectives.

guilpier-code commented on src/modeler/FileWriter.h:
```diff
@@ -54,8 +54,8 @@ class FileWriter: public Solver::IWriter
     explicit FileWriter(std::filesystem::path path);
 
-    void writeProblem(
-      const Optimisation::LinearProblemMpsolverImpl::OrtoolsLinearProblem& problem) override;
```
Why is class **FileWriter** for ? Do we need a class for what's done ?
- it has 2 member functions, doing different things, not consistent with one another, apart from "dumping".
- this class has 4 private data : 
  - **simulationId_** is shared by ... only one function (**writeSimulationTable(...)**), so what the point sharing this ?
  - **output** : `if(output)` is repeated in both functions. Then it should be used outside.
  - **studyPath_** : shouldn't be a member data because only used to build another data member.
- It seems the only reason why we put these 2 functions together in a class is because they share an output directory. So why not pass it as argument ?
- parent abstract class is useless : an interface class should not be created until we need it, that is if it has at least 2 children. So here it should be removed.

Instead, we could have : 
```c++
if (!parameters.noOutput)
{
	std::filesystem::path output_dir = createOuputDirectory(studyPath);
    Write(subproblem, "1-1.mps", output_dir);
	Write(master_problem, "master.mps", output_dir);
}
...
switch (solution->getStatus())
{
...
case MipStatus::FEASIBLE:
    WriteSimulationTable(...);
}
```

flomnes replied:
Yes, the idea of a `FileWriter` is quite nice but the actual design isn't very good

guilpier-code commented on src/modeler/modeler/Modeler.cpp:
```diff
@@ -144,38 +142,52 @@ void Modeler::run() const
     writer_.init(!parameters.noOutput, simulationTableSuffix);
-    writer_.writeProblem(ortools_linear_problem);
+    writer_.writeProblem(subproblem, "1-1.mps");
+    writer_.writeProblem(master_problem, "master.mps");
```
see [this comment ](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3211/files?show-viewed-files=true&file-filters%5B%5D=#r2494651820)

flomnes replied:
I introduced a free `Write` function

guilpier-code replied:
Please see [this (further) comment](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3211#discussion_r2499093906)

guilpier-code commented on src/tests/resources/modeler/13_1/input/model-libraries/lib_example_13_1.yml:
```diff
@@ -89,6 +89,7 @@ library:
           scenario-dependent: false
           time-dependent: false
       variables:
+        - id: master_only
```
do we want to keep this variable **master_only** ?

flomnes replied:
Yes ! This allowed us to spot a major bug with indices

flomnes committed:
Remove some functions

flomnes committed:
Filter

flomnes committed:
Rename

flomnes committed:
if

guilpier-code commented on src/optimisation/linear-problem-mpsolver-impl/include/antares/optimisation/linear-problem-mpsolver-impl/linearProblem.h:
```diff
@@ -79,11 +79,12 @@ class OrtoolsLinearProblem: public LinearProblemApi::ILinearProblem
     double infinity() const override;
     bool isLP() const override;
 
+    friend void Write(const OrtoolsLinearProblem& problem, const std::filesystem::path& path);
```
Here we still have dependency to an external **write** free function.
We should remove this friendship. 
This will cause a compilation error.
To make it go away,  we should make the following method **public** : 
```c++
operations_research::MPSolver* MpSolver() const;
```
and use it inside : 
```c++
void Write(const OrtoolsLinearProblem& problem, const std::filesystem::path& path)
```
If you hesitate to make **MpSolver()** public, don't : it **public** in class **LegacyOrtoolsLinearProblem**

guilpier-code commented on src/modeler/FileWriter.cpp:
Remove the boolean data member **output**, and the `if (output)` in class **FileWriter** : we don't need it if the client code uses the writer inside a : 
```c++
if (!parameters.noOutput)
{
    ...
}
```

guilpier-code commented on src/modeler/modeler/Modeler.cpp:
```diff
@@ -143,39 +142,60 @@ void Modeler::run() const
+    if (!parameters.noOutput)
+    {
+        writer_.init(!parameters.noOutput, simulationTableSuffix);
+        auto output = writer_.outputPath();
```
```c++
auto output = writer_.outputPath();
```
not intuitive : a writer returning a output path.
We should compose the output path outside the writer and pass it as argument to writer, for example at construction.
And remove the member function `outputPath()`

flomnes replied:
I agree but it's not blocking for the feature

guilpier-code committed:
6.2 c : fixes on PR 3190 (#3205)

Making fixes asked on [PR
3190](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3190)

---------

Co-authored-by: guilpier <guillaume.pierre_externe@rte-france.com>
Co-authored-by: Florian OMNES <florian.omnes@rte-france.com>

flomnes committed:
Remove bool FileWriter::output

flomnes committed:
Fix build
